src/gui/pedalboard/cables/Cable.cpp

```diff
@@ -112,3 +111,17 @@ void Cable::paint (Graphics& g)
         drawCable (g, startPortLocation, cableView.getCableMousePosition());
     }
 }
+
+Path Cable::pathGeneratorTask::createCablePath (juce::Point<float> start, juce::Point<float> end, float scaleFactor)
```

Comment:
Might need some help/explanation on generating this path with a separate thread

Reply:
For sure! So first off we probably don' want to use an `AudioUIBackgroundTask` since those are typically reserved for cases where we need the full buffer for whatever we want to do in the UI (like the oscilloscope). Here, probably just using a regular `juce::TimeSliceClient` should be fine.

Then all we need to do is override `TimeSliceClient::useTimeSlice()` to render the cable path(s), and register the client with a `TimeSliceThread`. It would be nice to have the cable drawers on the same `TimeSliceThread` as the Oscilloscope and the Tuner tasks; for that we could make a SharedResourcePointer ([like this](https://github.com/Chowdhury-DSP/chowdsp_utils/blob/master/modules/plugin/chowdsp_plugin_utils/Threads/chowdsp_AudioUIBackgroundTask.h#L36-L41)). Finally , we'll probably need to do some thread synchronization to make sure we're not introducing any data races or similar other threading issues.

One other thought is that we may want to have a single `TimeSliceClient` that renders the paths for all the Cables, rather than one per cable. Maybe it could live in `CableView` and also take care of calling `Cable::checkNeedsRepaint()` or equivalent. Like I'm kind of thinking of it like this:

Audio Thread:
- Generates level data and stores it in atomic variables

TimeSliceThread:
- Loops through all the cables:
  - Cable checks if its level has changed
  - If the level has changed, render a new cable path
  - Use `MessageManager::callAsync()` to trigger a repaint for the cable's bounds

Message Thread:
- Draw the new cable path (should happen automatically in `paint()` method)

Then there's two potential spots where threading issues could creep in, so we'll probably need to protect these spots with `juce::CriticalSection`s.
1. We need to make sure the `Cable` object isn't deleted (on the Message Thread), while being accessed on the TimeSliceThread.
2. We need to make sure the cable `Path` object isn't being simultaneously accessed on the Message and TimeSlice threads. We really don't want to block the MessageThread for long, so it's probably best to create the cable path using a temporary `Path` object, so then we only need to hold the `CriticalSection` while we `std::move()` the temporary Path into the Path object that will be rendered.

Reply:
Nice! This helps a lot thanks for the detailed write up. Using the `AudioUIBackgroundTask` didn't really make sense like you said so this clears things up nicely.